File: extensions/shared/src/main/java/app/morphe/extension/shared/spoof/js/nsigsolver/provider/JsChallengeProvider.java
package app.morphe.extension.shared.spoof.js.nsigsolver.provider;

import app.morphe.extension.shared.spoof.js.nsigsolver.common.CacheService;

import java.util.ArrayList;
import java.util.List;

public abstract class JsChallengeProvider {
    protected final CacheService cacheService = CacheService.getInstance();

    protected abstract List<JsChallengeType> getSupportedTypes();

    private void validateRequest(JsChallengeRequest request) throws JsChallengeProviderRejectedRequest {
        if (!getSupportedTypes().contains(request.getType())) {
            throw new JsChallengeProviderRejectedRequest(
                    "JS Challenge type " + request.getType() + " is not supported by the provider " + this.getClass().getSimpleName()
            );
        }
    }

    /**
     * Solve multiple JS challenges and return the results
     */
    public List<JsChallengeProviderResponse> bulkSolve(List<JsChallengeRequest> requests) {
        List<JsChallengeProviderResponse> responses = new ArrayList<>();
        List<JsChallengeRequest> validatedRequests = new ArrayList<>();

        for (JsChallengeRequest request : requests) {
            try {
                validateRequest(request);
                validatedRequests.add(request);
            } catch (JsChallengeProviderRejectedRequest e) {
                responses.add(new JsChallengeProviderResponse(request, e));
            }
        }

        // Add results from real implementation
        responses.addAll(realBulkSolve(validatedRequests));

        return responses;
    }

    /**
     * Subclasses can override this method to handle bulk solving
     */
    protected abstract List<JsChallengeProviderResponse> realBulkSolve(List<JsChallengeRequest> requests);
}

File: extensions/shared/src/main/java/com/liskovsoft/youtubeapi/app/nsigsolver/provider/JsChallengeProvider.kt
package com.liskovsoft.youtubeapi.app.nsigsolver.provider

import com.liskovsoft.youtubeapi.app.nsigsolver.common.YTInfoExtractor

internal abstract class JsChallengeProvider {
    protected abstract val supportedTypes: List<JsChallengeType>

    private fun validateRequest(request: JsChallengeRequest) {
        // Validate request using built-in settings
        if (request.type !in supportedTypes) {
            throw JsChallengeProviderRejectedRequest("JS Challenge type ${request.type} is not supported by the provider ${this::class.simpleName}")
        }
    }

    /**
     * Solve multiple JS challenges and return the results
     */
    fun bulkSolve(requests: List<JsChallengeRequest>): Sequence<JsChallengeProviderResponse> = sequence {
        val validatedRequests: MutableList<JsChallengeRequest> = mutableListOf()
        for (request in requests) {
            try {
                validateRequest(request)
                validatedRequests.add(request)
            } catch (e: JsChallengeProviderRejectedRequest) {
                yield(JsChallengeProviderResponse(request=request, error=e))
            }
        }
        yieldAll(realBulkSolve(validatedRequests))
    }

    /**
     * Subclasses can override this method to handle bulk solving
     */
    protected abstract fun realBulkSolve(validatedRequests: List<JsChallengeRequest>): Sequence<JsChallengeProviderResponse>

    protected fun getPlayer(videoId: String, playerUrl: String): String? {
        return try {
            YTInfoExtractor.loadPlayer(playerUrl)
        } catch (e: Exception) {
            throw JsChallengeProviderError("Failed to load player for JS challenge: $playerUrl", e)
        }
    }
}